### ISSUE TITLE
pgbackrest 2.12 breaks async archiving

### ISSUE THREAD

**Issue by jwpit** [pgbackrest/pgbackrest#711]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.12

2. PostgreSQL version: 10.7

3. Operating system/version: Debian Stretch

4. Did you install pgBackRest from source or from a package? 
apt.postgresql.org
5. Please attach the following as applicable:
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

Either broken pipe or process terminated unexpectedly on signal 11 errors  on differents cluster:
```
-------------------PROCESS START-------------------
2019-04-16 12:24:21.605 P00   INFO: archive-push-async command begin 2.12: [/mnt/data_san/10/adr/pg_wal] --cmd-ssh=/usr/bin/ssh --compress --compress-level=1 --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/mnt/data_san/10/adr --process-max=16 --repo1-host=10.101.0.4 --repo1-host-cmd=/usr/bin/pgbackrest --repo1-host-user=postgres --repo1-path=/adr/backup/svr_sql19/pgbackrest --stanza=adr
2019-04-16 12:24:21.614 P00   INFO: push 3134 WAL file(s) to archive: 000000020001B181000000AC...000000020001B18D000000EE
2019-04-16 12:24:23.728 P00  ERROR: [064]: unable to write to local-10 process write: [32] Broken pipe
```

```
-------------------PROCESS START-------------------
2019-04-16 12:26:27.981 P00   INFO: archive-push-async command begin 2.12: [/data/adrDL/pg_wal] --cmd-ssh=/usr/bin/ssh --compress --compress-level=1 --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/data/adrDL --process-max=16 --repo1-host=10.101.0.4 --repo1-host-cmd=/usr/bin/pgbackrest --repo1-host-user=postgres --repo1-path=/adr/backup/svr_sql19/pgbackrest --stanza=adrdl
2019-04-16 12:26:27.981 P00   INFO: push 5 WAL file(s) to archive: 0000000300002FD2000000AC...0000000300002FD2000000B0
2019-04-16 12:26:28.481 P00   INFO: archive-push-async command end: aborted with exception [102]
2019-04-16 12:26:29.639 P00  ERROR: [102]: local-11 process terminated unexpectedly on signal 11
2019-04-16 12:26:31.545 P00   INFO: archive-push-async command end: aborted with exception [102]
postgres@svr-psql19:~$ tail /var/log/pgbackrest/adrpb-archive-push-async.log
2019-04-16 12:27:30.882 P00   INFO: push 5 WAL file(s) to archive: 00000003000000DE00000048...00000003000000DE0000004C
2019-04-16 12:27:31.413 P00   INFO: archive-push-async command end: aborted with exception [102]
2019-04-16 12:27:32.483 P00  ERROR: [102]: local-11 process terminated unexpectedly on signal 11

-------------------PROCESS START-------------------
2019-04-16 12:27:33.898 P00   INFO: archive-push-async command begin 2.12: [/data/adrPB/pg_wal] --cmd-ssh=/usr/bin/ssh --compress --compress-level=1 --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/data/adrPB --process-max=16 --repo1-host=10.101.0.4 --repo1-host-cmd=/usr/bin/pgbackrest --repo1-host-user=postgres --repo1-path=/adr/backup/svr_sql19/pgbackrest --stanza=adrpb
2019-04-16 12:27:33.898 P00   INFO: push 5 WAL file(s) to archive: 00000003000000DE00000048...00000003000000DE0000004C
2019-04-16 12:27:34.389 P00   INFO: archive-push-async command end: aborted with exception [102]
2019-04-16 12:27:35.495 P00  ERROR: [102]: local-11 process terminated unexpectedly on signal 11
2019-04-16 12:27:37.401 P00   INFO: archive-push-async command end: aborted with exception [102]
```

syslog:
```
Apr 16 12:24:23 svr-psql19 kernel: [7670279.213166] pgbackrest[33786]: segfault at 351 ip 0000559b0beb82a7 sp 00007ffd7299c9b0 error 4 in pgbackrest[559b0be65000+f800
0]
Apr 16 12:24:23 svr-psql19 kernel: [7670279.213514] pgbackrest[33792]: segfault at 30 ip 000055abf4b462a7 sp 00007ffdc43b40f0 error 4 in pgbackrest[55abf4af3000+f8000
]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.549589] show_signal_msg: 6 callbacks suppressed
Apr 16 12:24:28 svr-psql19 kernel: [7670283.549592] pgbackrest[34007]: segfault at 30 ip 00005634c3dc8641 sp 00007ffdcf05e3a0 error 4 in pgbackrest[5634c3d92000+f8000
]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.558528] traps: pgbackrest[34003] general protection ip:55e2c908eab1 sp:7ffe8f05b260 error:0 in pgbackrest[55e2c9060000+f80
00]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.561730] traps: pgbackrest[33993] general protection ip:55669530daf2 sp:7ffe2380ffc0 error:0 in pgbackrest[5566952df000+f80
00]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.563145] pgbackrest[34008]: segfault at 30 ip 00005633c32312a7 sp 00007ffc552c9680 error 4 in pgbackrest[5633c31de000+f8000
]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.565235] pgbackrest[33989]: segfault at 351 ip 00005621420212a7 sp 00007ffd81d1eb60 error 4 in pgbackrest[562141fce000+f800
0]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.565657] traps: pgbackrest[34012] general protection ip:55ef2a841ab1 sp:7ffe79ae4980 error:0 in pgbackrest[55ef2a813000+f80
00]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.567780] pgbackrest[33998]: segfault at 30 ip 000055aa7cf132a7 sp 00007fffc1d3d0d0 error 4 in pgbackrest[55aa7cec0000+f8000
]
Apr 16 12:24:28 svr-psql19 kernel: [7670283.602616] traps: pgbackrest[34002] general protection ip:55910d512ab1 sp:7ffcaa2b1880 error:0 in pgbackrest[55910d4e4000+f80
00]

```

7. Describe the issue:
Upgrading to 2.12 broke the async WAL archiving.
I had to downgrade to 2.11 to get it working again.

**Comment by dwsteele** [comment]
It looks like we have a solution for the broken pipe errors, but we have not been able to reproduce any seg faults -- unless they are related somehow.

**Comment by dwsteele** [comment]
The broken pipe error has been fixed at b960919cf7322f0 and released in 2.13.  Hopefully this will allow us to see what's actually going on here -- if it's not the same issue as #712.

Let me know if you would like to appear in the release notes with a different name than `jwpit`.

**Comment by dwsteele** [comment]
Are you still seeing these errors in 2.13?

**Comment by jwpit** [comment]
> Are you still seeing these errors in 2.13?

Yes, unfortunately:

Apr 25 16:18:18 svr-psql19 kernel: [8461935.796989] traps: pgbackrest
[89125] general protection ip:560f5ff28fd2 sp:7fff8565f9c0 error:0 in
pgbackrest[560f5fef5000+ff000]
Apr 25 16:18:22 svr-psql19 kernel: [8461940.003706] pgbackrest[89486]:
segfault at 351 ip 000056237f842ba7 sp 00007ffd0e0ec050 error 4 in
pgbackrest[56237f7ea000+ff000]

**Comment by cmwshang** [comment]
@jwpit is it possible for you to run addr2line with 0x58BA7 (e.g. 0x000056237f842ba7 – 0x56237f7ea000 = 0x58BA7)

**Comment by jwpit** [comment]
Cynthia Shang <notifications@github.com> wrote on 04/25/2019 05:34:25 PM:

> @jwpit is it possible for you to run addr2line with 0x58BA7 (e.g.
> 0x000056237f842ba7 – 0x56237f7ea000 = 0x58BA7)


Not sure about the syntax:

addr2line  -e /usr/bin/pgbackrest 0x58BA7
??:?

**Comment by cmwshang** [comment]
@jwpit Hmmm. Try adding -fCi (addr2line -e /usr/bin/pgbackrest -fCi 0x58BA7).

**Comment by jwpit** [comment]
Cynthia Shang <notifications@github.com> wrote on 04/25/2019 06:04:57 PM:

> @jwpit Hmmm. Try adding -fCi (addr2line -e /usr/bin/pgbackrest -fCi
0x58BA7).

 addr2line  -e /usr/bin/pgbackrest -fCi 0x58BA7
storageDriverRemoteNew
??:?

**Comment by cmwshang** [comment]
This looks similar to this issue: https://github.com/pgbackrest/pgbackrest/issues/608. The end result was that they had a perl version conflict - see some of the comments on that issue from Oct 8 and above. Maybe a conflict is causing your issue as well?

**Comment by jwpit** [comment]
Both Systems have similar Perl installations.
That doesn't look like a version conflict.

# dpkg -l perl*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                Version                Architecture           Description
+++-===================================-======================-======================-============================================================================
ii  perl                                5.24.1-3+deb9u5        amd64                  Larry Wall's Practical Extraction and Report Language
ii  perl-base                           5.24.1-3+deb9u5        amd64                  minimal Perl system
un  perl-cross-config                   <none>                 <none>                 (no description available)
un  perl-dbdabi-94                      <none>                 <none>                 (no description available)
un  perl-doc                            <none>                 <none>                 (no description available)
un  perl-modules                        <none>                 <none>                 (no description available)
ii  perl-modules-5.24                   5.24.1-3+deb9u5        all                    Core Perl modules
un  perl-openssl-abi-1.1                <none>                 <none>                 (no description available)
ii  perl-openssl-defaults:amd64         3                      amd64                  version compatibility baseline for Perl OpenSSL packages

**Comment by jwpit** [comment]
There are still broken pipe Errors:

-------------------PROCESS START-------------------
2019-04-26 11:10:27.960 P00   INFO: archive-push-async command begin 2.13: [/mnt/data_san/10/adr/pg_wal] --cmd-ssh=/usr/bin/ssh --compress --compress-level=1 --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/mnt/data_san/10/adr --process-max=16 --repo1-host=10.101.0.4 --repo1-host-cmd=/usr/bin/pgbackrest --repo1-host-user=postgres --repo1-path=/adr/backup/svr_sql19/pgbackrest --stanza=adr
2019-04-26 11:10:27.962 P00   INFO: push 30 WAL file(s) to archive: 000000020001B3C800000093...000000020001B3C8000000B0
2019-04-26 11:10:29.736 P00  ERROR: [102]: local-10 process terminated unexpectedly on signal 11

-------------------PROCESS START-------------------
2019-04-26 11:10:32.204 P00   INFO: archive-push-async command begin 2.13: [/mnt/data_san/10/adr/pg_wal] --cmd-ssh=/usr/bin/ssh --compress --compress-level=1 --log-level-console=off --log-level-file=info --log-level-stderr=off --pg1-path=/mnt/data_san/10/adr --process-max=16 --repo1-host=10.101.0.4 --repo1-host-cmd=/usr/bin/pgbackrest --repo1-host-user=postgres --repo1-path=/adr/backup/svr_sql19/pgbackrest --stanza=adr
2019-04-26 11:10:32.206 P00   INFO: push 28 WAL file(s) to archive: 000000020001B3C800000093...000000020001B3C8000000B0
2019-04-26 11:10:34.032 P00  ERROR: [064]: unable to write to local-16 process write: [32] Broken pipe

**Comment by dwsteele** [comment]
First I would try disabling async-archive and see what happens.  If that fails increase console-log-level to trace and send that log.

If it succeeds then we'll need trace logs from the async command with `--log-subprocess` set.

BTW - `process-max=16` is really high for archive get.  2-4 is usually fine.  You may be hitting some resource limits.

**Comment by jwpit** [comment]
async-archive=n works.
I sent you the tracelog by mail.
setting --process-max=4 with async-archive=y works, too

**Comment by dwsteele** [comment]
Are you keeping up OK with 4 processes or are you queuing a lot?

We'd like to see the log for the local process that threw the error so we can see what it was doing.  Should be adr-archive-push-async-local-009.log

**Comment by dwsteele** [comment]
OK, thanks, that log led me straight to the bug.  Should be OK to set `process-max=8` if you need to, but sounds like you are getting along fine with 4.

**Comment by dwsteele** [comment]
Fix committed at d0c296bd5b679.

This doesn't seem to require a release since process-max > 8 is unlikely to be useful.  Let me know if you think otherwise.
